warcprox/mitmproxy.py

```diff
@@ -439,12 +450,17 @@ def _proxy_request(self, extra_response_headers={}):
                             self._max_resource_size, self.url)
                         break
 
+            connection_is_fine = True
             self.log_request(prox_rec_res.status, prox_rec_res.recorder.len)
         finally:
-            # Let's close off the remote end
+            # Let's close off the remote end. If remote connection is fine,
+            # put it back in the pool to reuse it later.
             if prox_rec_res:
                 prox_rec_res.close()
-            self._remote_server_sock.close()
+            if connection_is_fine and not is_connection_dropped(self._remote_server_conn):
```

Comment:
This point has to do with the Python 3.4 test failing. I figured out that connection reusing has to do with this issue.

If we comment out this part of the code that does `self._conn_pool._put_conn(self._remote_server_conn)`, all tests pass.
Using just the `connection_is_fine` condition, made all tests fail.
The extra `is_connection_dropped` check https://github.com/shazow/urllib3/blob/master/urllib3/util/connection.py#L7
improves the situation but still one test fails in Python 3.4.

@nlevitt maybe you have an idea on how to fix the test?